### ISSUE TITLE
Vllm_importance_sampling_correction sequence-level mode aggregates per-token log-ratios with sum instead of mean

### ISSUE THREAD

**Issue by ionut-anghelina** [huggingface/trl#5814]
### Reproduction

  In `trl/trainer/grpo_trainer.py`, the `vllm_importance_sampling_correction` block aggregates per-token log-probability differences with **sum** instead of **mean** when
  `vllm_importance_sampling_mode` is `sequence_mask` or `sequence_truncate`. The same file uses **mean** for the policy IS at `importance_sampling_level="sequence"` (line ~2501).
  This inconsistency produces an IS ratio whose variance scales with completion length, and combined with `clamp(max=cap)` it deflates the effective gradient even when per-token
  logprobs from vLLM and the trainer agree.

Bug location: 
  ```python
  # trl/trainer/grpo_trainer.py, line ~2070 (main as of 2026-05-22)
  sequence_level_is = self.vllm_importance_sampling_mode in ["sequence_mask", "sequence_truncate"]
  if sequence_level_is:
      per_sequence_logps_diff = per_token_logps_diff.sum(dim=-1, keepdim=True)   # ← bug
      logps_diff = per_sequence_logps_diff
  ```

## Expected vs Actual

  - **Expected**: with `num_iterations=1` (vLLM-policy = trainer-policy at gradient time) and matching per-token logprobs, the IS ratio mean should be ≈ 1.0 with tight spread.
  - **Actual**: IS ratio mean lands at **0.4–0.7** with wide spread (0.0 to cap), even when `sampling/sampling_logp_difference/mean` is near zero.

This is a numerical bug in the aggregation, so a torch-only synthetic reproducer is exact:

  ```python
  import torch

  torch.manual_seed(0)
  B, T, cap = 16, 200, 3.0  # 16 sequences, 200 completion tokens, default cap
  mask = torch.ones(B, T)
  # Realistic numerical noise between vLLM and trainer recompute: mean=0, std≈0.1
  per_token_logps_diff = torch.randn(B, T) * 0.1 * mask

  # Current TRL behavior (trainer/grpo_trainer.py:2070)
  seq_diff_sum  = per_token_logps_diff.sum(dim=-1, keepdim=True)
  ratio_sum     = torch.exp(seq_diff_sum).clamp(max=cap)

  # Proposed fix (mean — matches the existing policy-IS aggregation at line ~2501)
  seq_len       = mask.sum(dim=-1, keepdim=True).clamp(min=1.0)
  seq_diff_mean = per_token_logps_diff.sum(dim=-1, keepdim=True) / seq_len
  ratio_mean    = torch.exp(seq_diff_mean).clamp(max=cap)

  print(f"sum-based  IS ratio: mean={ratio_sum.mean():.3f}  std={ratio_sum.std():.3f}  "
        f"min={ratio_sum.min():.3f}  max={ratio_sum.max():.3f}")
  print(f"mean-based IS ratio: mean={ratio_mean.mean():.3f}  std={ratio_mean.std():.3f}  "
        f"min={ratio_mean.min():.3f}  max={ratio_mean.max():.3f}")
  ```

  Output:

  ```
  sum-based  IS ratio: mean=1.286  std=1.166  min=0.121  max=3.000
  mean-based IS ratio: mean=0.999  std=0.008  min=0.990  max=1.014
  ```

### System Info

  - **TRL version:** 1.4.0 
  - **Transformers:** 5.8.0
  - **PyTorch:** 2.10.0 + CUDA 12.8
  - **vLLM:** 0.18.0 

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by qgallouedec** [comment]
Thanks for the careful writeup and the follow-up PR but I don't think this is a bug. I think the conclusion comes from conflating two things that share the word "sequence" in this file but are different mathematical objects.

### The two aggregations are not the same object

`vllm_importance_sampling_mode="sequence_*"` is the **off-policy correction** between vLLM and the trainer. The mathematically correct sequence-level IS weight here is the actual sequence probability ratio:

$$
\rho_{\text{seq}}(y)
= \frac{\pi_{\text{train}}(y \mid x)}{\pi_{\text{vllm}}(y \mid x)}
= \prod_{t=1}^{\lvert y \rvert}
\frac{\pi_{\text{train}}(y_t \mid x, y_{ < t})}{\pi_{\text{vllm}}(y_t \mid x, y_{ < t})}
= \exp\left(
\sum_{t=1}^{\lvert y \rvert}
\log
\frac{\pi_{\text{train}}(y_t \mid x, y_{ < t })}
{\pi_{\text{vllm}}(y_t \mid x, y_{ < t})}
\right).
$$

That's exactly the unbiased estimator written down in our own docs ([`paper_index.md` § "Sequence-level Importance Sampling"](https://huggingface.co/docs/trl/en/paper_index#sequence-level-importance-sampling)) *"the correct, unbiased policy gradient estimator applies a single importance ratio over the entire generated sequence"*, and drawn from [*"When Speed Kills Stability"*](https://yingru.notion.site/When-Speed-Kills-Stability-Demystifying-RL-Collapse-from-the-Training-Inference-Mismatch-271211a558b7808d8b12d403fd15edda) §4.2.1. It's the explicit design rationale for the `sequence_*` modes added in #4530 (see #4493). [TIS](https://fengyao.notion.site/off-policy-rl) and [MIS](https://ringtech.notion.site/icepop) sit on top of this ratio as the variance-reduction mechanism, with `vllm_importance_sampling_cap` as the knob.

`importance_sampling_level="sequence"` is the **GSPO policy IS weight** inside the PPO-clipped loss. [GSPO](https://huggingface.co/papers/2507.18071) §4.1 eq. (7) defines its sequence ratio as the *geometric mean*:

$$
s_i(\theta) = \left( \frac{\pi_\theta(y_i \mid x)}{\pi_{\theta_{\text{old}}}(y_i \mid x)} \right)^{\frac{1}{\lvert y_i \rvert}} = \exp\left( \frac{1}{\lvert y_i \rvert} \sum_{t=1}^{\lvert y_i \rvert} \log \frac{\pi_\theta(y_{i,t} \mid \cdot)}{\pi_{\theta_{\text{old}}}(y_{i,t} \mid \cdot)} \right).
$$

The paper itself states the $1/|y_i|$ is there *"to reduce the variance and to control $s_i(\theta)$ within a unified numerical range"* so the clip range $[1-\varepsilon, 1+\varepsilon]$ stays meaningful across sequence lengths. That's the same intuition you're invoking, but it applies to the **policy** ratio used inside the clipped objective, not to the off-policy correction between two engines.

### Why "sum → mean" would be wrong

Aligning L2069 to use the mean would turn an unbiased off-policy correction into a biased geometric mean, i.e. it would silently double-apply the GSPO length-normalization on top of the off-policy correction, and pull the ratio toward 1 for *any* per-token mismatch. That's why your "$0.569 \to 0.993$" empirical signal in #5815 doesn't actually distinguish "the aggregation was wrong" from "the unbiased estimator is correctly reflecting a vLLM↔trainer logp mismatch": the geometric mean is *guaranteed* to live nearer 1 than the product, regardless of which hypothesis is true.

### On the synthetic reproducer

The reproducer's "expected $\approx 1.0$ with tight spread" already assumes the geometric-mean answer. Under your noise model ($\sigma = 0.1$ per-token, $T = 200$), an *unbiased* sequence IS ratio is lognormal with

$$
\mathbb{E}\left[ \exp\left( \sum_{t=1}^{T} X_t \right) \right] = \exp\left( \frac{T \sigma^2}{2} \right) = \exp\left( \frac{200 \cdot 0.01}{2} \right) \approx 2.72.
$$

So $\text{mean}=1.286,\ \text{std}=1.166$ is the correct behavior of an unbiased lognormal estimator (the discrepancy from $2.72$ is because the cap clips the right tail). That's exactly the variance the cap exists to absorb.

---

The shared word "sequence" in both config flags is admittedly a bit misleading, but I don't think there's a clean rename: each flag individually describes a real token-vs-sequence granularity choice, just applied to different ratios serving different purposes.

Closing as not-a-bug, and I'll close #5815 with a pointer here; happy to reopen if you have a concrete training-run signal that the sum aggregation is wrong (rather than that the underlying mismatch is large, which is a real and separate problem).
